### ISSUE TITLE
gnatsd-server sometimes TIME_WAIT

### ISSUE THREAD

**Issue by xiaoshengaimm** [nats-io/gnatsd#449]
- [ ] gnatsd-server tcp connection sometimes all time_wait，that all client do not reconnect;
 - [ ] gnatsd-server verion is 0.7.2; upgrade to 0.8.0 happen again too.

 
## Defects 

TCP connection keep TIME_WAIT and all message can not publish.

until restart the gnatsd-server, all client reconnect and the connections return to ESTABLISHED.

no messages find on logs ...... .....

this happens twice on different server.

server is centos 6.6; 32 cpus and 48G; when this happening the server is ok..

![image](https://cloud.githubusercontent.com/assets/1866149/23933337/d046511e-0978-11e7-8dc9-88b5292b5eed.png)

**Comment by derekcollison** [comment]
Run the server with monitoring and please add what http://$gnatsdhost/varz reports. You might be running out of connections. What does ulimit -a report from where you run the server?

**Comment by xiaoshengaimm** [comment]
second time on this morning

![image](https://cloud.githubusercontent.com/assets/1866149/26661591/3bc7e478-46b1-11e7-99f4-f7e55263a608.png)

**Comment by xiaoshengaimm** [comment]
second time this morning .

![image](https://cloud.githubusercontent.com/assets/1866149/26661619/5d506930-46b1-11e7-96c5-a8e5c16c56a7.png)

**Comment by derekcollison** [comment]
Could you run against the latest released binary, which is 0.9.6.

https://github.com/nats-io/gnatsd/releases/tag/v0.9.6

Also, can you tell us more about the clients? Which language? Which version? And do the clients seem to be disconnected all at once?

Thanks,
=derek

**Comment by xiaoshengaimm** [comment]
while i try 0.9.6，the same problem happen

client is ruby-nats, 0.5.1; the clients disconnected all at once...

**Comment by wallyqs** [comment]
hi @xiaoshengaimm which version of Ruby is it? Latest release of ruby-nats is [v0.8.2](https://github.com/nats-io/ruby-nats/releases/tag/v0.8.2) so would recommend upgrading as well. The client has `on_reconnect`, `on_disconnect`, `on_close`, `on_error` [callbacks](https://github.com/nats-io/ruby-nats#clustered-usage) as well which might help in debugging further the issue in case all the clients are Ruby based.

**Comment by xiaoshengaimm** [comment]
because there are thousands of clients running online, so update the ruby-nats version is very hard to us.
i print debug info on the gnatsd server，all client connection is print "Client Ping Timer" and never send message again...

**Comment by kozlovic** [comment]
You say that you see "Client Ping Timer" in debug mode in the server's log. Do you see "Stale Client Connection - Closing" for all your clients?
I assume that the clients are not hosted on the same machine than the server, and maybe running on different hosts. It could be that the host machine where NATS Server is running is suddenly cut off the network, the network card may have issues, etc... If clients are distributed and they are all disconnecting from the server at the same time, this seems to be the most probable cause.

**Comment by xiaoshengaimm** [comment]
i may cause by route connection reset...but why the route connection closed...and route connection reset cause this problem?
![image](https://user-images.githubusercontent.com/1866149/27983457-1cb549e4-63ef-11e7-908a-a15f005c1aa9.png)
![image](https://user-images.githubusercontent.com/1866149/27983472-7fc2b5f8-63ef-11e7-8aa0-70e5bc4af710.png)
![image](https://user-images.githubusercontent.com/1866149/27983474-8b35954a-63ef-11e7-9f7c-c7f1c380b44e.png)

**Comment by xiaoshengaimm** [comment]
At last, I change nats-server back to ruby nats-server 0.5.1, it is stable now....

gnatsd cluster mode can not be stable. one of cluster server change to "Slow Consumer"，route connection disconnect. howerver, after route connection reconnect, messages never route to the "Slow Consumer", it have to kill the Slow Consumer process and restart, the cluster mode return to work....

hope your test. Thanks.

**Comment by kozlovic** [comment]
I did test and cannot reproduce. Is the `Slow Consumer` coming from the route itself (should be `rid:xxx - Slow Consumer`, or from a client connection: `cid:xxx - Slow Consumer`). Even when artificially slowing down the route reader loop to cause the upstream server to disconnect, the connection is later re-established and the messages flow again.
Note that in above screen shots, I see a lot of `Bad or Missing ROUTER Identity` debug messages. There was a fix in 0.7.0 that would prevent this from happening, so not sure what version was used in outputs above.. unless, you have a client connecting to the cluster port, in which case, from the log it may look like a route (`rid`, etc..) but actually be a client, and for a consumer, this would produce the said debug outputs.
In the upcoming `1.0.0` server release (planned today/tomorrow) the server will reject a client connecting to a cluster port. Here is the output of a subscriber connecting to wrong port with current 0.9.6 release:
```
[8010] 2017/07/12 09:46:34.355696 [DBG] 127.0.0.1:56287 - rid:1 - Route connection created
[8010] 2017/07/12 09:46:44.121466 [DBG] ::1:56293 - cid:2 - Client connection created
[8010] 2017/07/12 09:46:44.121895 [DBG] ::1:56293 - cid:2 - Bad or Missing ROUTER Identity, not processing msg
[8010] 2017/07/12 09:46:44.122930 [DBG] ::1:56293 - cid:2 - Client connection closed
```
The first line is the subscriber connecting to the wrong port: server says that a route is connected. After that, a publisher connects to the client port and sends a message. Notice the debug messages.

This is the output of the same test with current master branch or upcoming `1.0.0` executable:
```
[8064] 2017/07/12 09:48:44.404257 [DBG] 127.0.0.1:56322 - rid:1 - Route connection created
[8064] 2017/07/12 09:48:44.404673 [DBG] 127.0.0.1:56322 - rid:1 - Router connection closed
[8064] 2017/07/12 09:48:44.404704 [ERR] 127.0.0.1:56322 - rid:1 - Error reading from client: Attempted To Connect To Route Port
```
As you can see, the server rejects that connection.

**Comment by kozlovic** [comment]
Closing due to age.
